### PR TITLE
`redo what` tab completion feature

### PR DESCRIPTION
**Pull request by MyApaulogies**
Create the redo command `redo what_predefined`, which displays the commands listed in `redo what` which will always be listed regardless of the subdirectory of the project you're in.

### CODE REVIEW

**Comment by MyApaulogies** [comment]
Everything should be good. There are a few edge cases, like `cd`ing into a `build` directory still brings up completions, and trying to complete `redo ../adamant` while in `~/adamant` doesn't work, and also it's a bit slower than it could be (rewrite it in Rust?) It seems usable now though, if someone wants to test it out.

**Comment by MyApaulogies** [comment]
Looks like all the failing tests are typos in files I didn't change...

**Comment by MyApaulogies** [comment]
> [dinkelk](https://github.com/dinkelk) and others added 18 commits May 30, 2024 23:26
Umm, any idea what happened here? In any case, I cleaned up my commit history, so it should be ready for a merge if you think it's good.

**Comment by Jbsco** [comment]
Try using `rebase -i -r HEAD~#` to drop commits not authored in this PR, then rebasing on main, then again with `--committer-date-is-author-date` to restore inline commit dates. There may be other ways to do this, but should work.

**Review by Jbsco — APPROVED** [review]
`rebase -i -r` and drop of commits not authored in this PR seems to have done the trick!

File: redo/rules/build_what_predefined.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        "prove",
+        "analyze",
+        "style",
+        "pretty",
+        "test_all",
+        "coverage_all",
+        "style_all",
+    ]
+    # with open('../predefined_targets.txt') as f:
+    #     return f.read().split()
```
Comment (dinkelk) [inline redo/rules/build_what_predefined.py:27]: Remove commented out code

File: redo/rules/build_what_predefined.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+class build_what_predefined(build_rule_base):
+    '''Lists all redo targets which will be available in any subdirectory
+    of the project base folder. See `build_what`.
+    '''
+
+    # similar to redo what, no need to access database
+    def build(self, redo_1, redo_2, redo_3):
+        # directory = os.path.dirname(os.path.abspath(redo_1))
+        # environ["BUILD_PATH"] = directory + os.pathsep + directory + os.sep + ".."
+        # super(build_what_predefined, self).build(redo_1, redo_2, redo_3)
```
Comment (dinkelk) [inline redo/rules/build_what_predefined.py:39]: remove commented out code.

File: env/redo_completion.sh [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    local cmdname=$1
+    local this_word=$2
+    local prev_word=$3
+
+    # echo
+    # echo cmdname = $cmdname
+    # echo this_word = $this_word
+    # echo prev_word = $prev_word
+    # echo first_run = $first_run
```
Comment (dinkelk) [inline env/redo_completion.sh]: Remove commented out code

File: env/redo_completion.sh [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            # pid is not alive, so start bg process
+            (__do_caching >/dev/null &)
+            __redo_what_bg_pid=$!
+        fi
+
+    fi
+
+
+
+
```
Comment (dinkelk) [inline env/redo_completion.sh:420]: spacing

**Review by dinkelk — CHANGES_REQUESTED** [review]
I think some cleanup needs to happen in env/redo_completion.sh. There is a lot of unnecessary white space and commented out code that hinders readability. The actual code looks good though.

**Comment by MyApaulogies** [comment]
Now it should be done, I took your suggestions and made a few other changes. (but the of course tests are failing for some reason...)

**Comment by MyApaulogies** [comment]
It looks like they're all permissions errors... Does this have to do with changing the Dockerfile to end with `USER user` instead of `USER root`?

**Review by dinkelk — APPROVED** [review]

**Pull request merged**